uefi/firmwarevolume.go

```diff
@@ -189,6 +195,36 @@ func (fv *FirmwareVolume) Assemble() ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Make sure we don't go over the size.
+	fvLen := uint64(len(fv.buf))
+	fOffset := fv.DataOffset
+	for _, f := range fv.Files {
+		fBuf, err := f.Assemble()
+		if err != nil {
+			return nil, err
+		}
+		fLen := uint64(len(fBuf))
+		// We have to pad to the 8 byte alignments.
+		alignedOffset := align8(fOffset)
+		if alignedOffset > fvLen {
+			return nil, fmt.Errorf("insufficient space in %#x bytes FV, files too big", fvLen)
+		}
+		fillFFs(fv.buf[fOffset:alignedOffset])
+		fOffset = alignedOffset
+
+		// Check size
+		if fLen+fOffset > fvLen {
+			// TODO: Actually loop through and calculate the full size so we know how much to reduce by.
+			// For now we just return early
+			return nil, fmt.Errorf("insufficient space in %#x bytes FV, files too big", fvLen)
+		}
+		// Overwrite old data in the firmware volume.
+		copy(fv.buf[fOffset:], fBuf)
+		fOffset += fLen
+	}
+
+	// Fill to the end with FFs
+	fillFFs(fv.buf[fOffset:fvLen])
```

Comment:
I'm sorry, because of the last refactor, you actually need an if-statement here now:

```
if fOffset < fvLen {
    fillFFs(fv.buf[fOffset:fvLen])
}
```

Reply:
Done